src/mat/impls/inv/matinv.c

```diff
@@ -546,33 +551,33 @@ static PetscErrorCode MatInvCreateInnerObjects_Inv(Mat imat)
     }
   }
 
-  if (inv->type == MAT_INV_BLOCKDIAG) {
-    KSP *kspp;
-    TRY( KSPSetType(inv->ksp, KSPPREONLY) );
-    TRY( KSPGetPC(inv->ksp, &pc) );
-    TRY( PCSetType(pc, PCBJACOBI) );
-    TRY( PCSetUp(pc) );
-    TRY( PCBJacobiGetSubKSP(pc, PETSC_IGNORE, PETSC_IGNORE, &kspp) );
-    inv->innerksp = *kspp;
-  } else if (inv->redundancy) {
+  if (inv->type == MAT_INV_BLOCKDIAG || inv->redundancy) {
     const char *prefix;
-    char stri[1024];
     TRY( KSPSetType(inv->ksp, KSPPREONLY) );
     TRY( KSPGetPC(inv->ksp, &pc) );
     TRY( MatGetOptionsPrefix(imat,&prefix) );
     TRY( PCSetOptionsPrefix(pc,prefix) );
     TRY( PCAppendOptionsPrefix(pc,"mat_inv_") );
-    TRY( PetscSNPrintf(stri, sizeof(stri), "-%smat_inv_redundant_pc_type none",prefix) );
-    TRY( PetscOptionsInsertString(NULL,stri) );
-    //petsc bug start https://bitbucket.org/petsc/petsc/branch/hzhang/fix-pcredundant#diff
-    TRY( PetscSNPrintf(stri, sizeof(stri), "-psubcomm_type %s", PetscSubcommTypes[inv->psubcommType]) );
-    TRY( PetscOptionsInsertString(NULL,stri) );
-    TRY( PCSetFromOptions(pc) );
-    //bug end
-    TRY( PCSetType(pc, PCREDUNDANT) );
-    TRY( PCRedundantSetNumber(pc, inv->redundancy) );
-    TRY( PCSetUp(pc) );//not necessary after fix, see above
-    TRY( PCRedundantGetKSP(pc, &inv->innerksp) );
+    if (inv->type == MAT_INV_BLOCKDIAG) {
+      KSP *kspp;
+      TRY( PCSetType(pc, PCBJACOBI) );
+      TRY( PCSetUp(pc) );
+      TRY( PCBJacobiGetSubKSP(pc, PETSC_IGNORE, PETSC_IGNORE, &kspp) );
+      inv->innerksp = *kspp;
+    } else {
+      char stri[1024];
+      TRY( PetscSNPrintf(stri, sizeof(stri), "-%smat_inv_redundant_pc_type none",prefix) );
+      TRY( PetscOptionsInsertString(NULL,stri) );
+      //petsc bug start https://bitbucket.org/petsc/petsc/branch/hzhang/fix-pcredundant#diff
+      TRY( PetscSNPrintf(stri, sizeof(stri), "-psubcomm_type %s", PetscSubcommTypes[inv->psubcommType]) );
+      TRY( PetscOptionsInsertString(NULL,stri) );
+      TRY( PCSetFromOptions(pc) );
+      //bug end
+      TRY( PCSetType(pc, PCREDUNDANT) );
+      TRY( PCRedundantSetNumber(pc, inv->redundancy) );
+      TRY( PCSetUp(pc) );//not necessary after fix, see above
```

Comment:
... and this?

Reply:
that was already fixed in 07d0622

Reply:
aha, ok, sorry - I was looking at the old commit

Reply:
Wait - you just removed the comments, right? I was thinking that the comments say the ugly code is a workaround that could be revisited once that branch gets merged. It was this https://gitlab.com/petsc/petsc/-/commit/f07d178c4121a870076ccb0b7ed834dcdbc01e49

No need to deal with that right away but perhaps we could keep the comments and create an issue. I or you should take a look at it at some point.

Reply:
OK, issue #53 created. No need to revert the comments since this discussion and the issue does the job. Resolving.

Reply:
I actually checked before removing the comments ;) And simply removing `PCSetUp` makes problems elsewhere. However, looking at it again I can save a few lines of code (saving a NOP `PcSetUp`). See the amended last commit.